README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# donut

[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/paulnorthrop/donut?branch=master&svg=true)](https://ci.appveyor.com/project/paulnorthrop/donut)
[![R-CMD-check](https://github.com/paulnorthrop/donut/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/paulnorthrop/donut/actions/workflows/R-CMD-check.yaml)
[![Coverage
Status](https://codecov.io/github/paulnorthrop/donut/coverage.svg?branch=master)](https://app.codecov.io/github/paulnorthrop/donut?branch=master)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/donut)](https://cran.r-project.org/package=donut)
[![Downloads
(monthly)](https://cranlogs.r-pkg.org/badges/donut?color=brightgreen)](https://cran.r-project.org/package=donut)
[![Downloads
(total)](https://cranlogs.r-pkg.org/badges/grand-total/donut?color=brightgreen)](https://cran.r-project.org/package=donut)

## Nearest Neighbour Search with Variables on a Torus

### What does donut do?

There are several R packages, such as
[RANN](https://cran.r-project.org/package=RANN) and
[nabor](https://cran.r-project.org/package=nabor) that find the $k$
nearest neighbours in a dataset of specified query points, based on some
metric, such as L2 or L1. The donut package considers the situation
where one or more of the variables in the dataset is periodic on a
finite interval. For example, direction is periodic on the interval
$(0, 360)$ degrees. In the small dataset $\{10, 90, 350\}$ degrees 350
is closer to 10 than is 90: 10 and 350 are separated by 20 degrees, 10
and 90 by 80 degrees.

The function `nnt()` finds the $k$ nearest neighbours of each of a set
of points of interest, wrapping periodic variables on a torus so that
this periodicity is reflected. The user chooses the function to use to
find the nearest neighbours. The nearest neighbour functions from the
aforementioned packages are used as examples.

### An example

We use a simple example from the `RANN:nn2()` documentation. We suppose
that both variables should be wrapped, on the ranges $(0, 2\pi)$ and
$(0, 3)$ respectively. We choose the query points of interest to
illustrate the wrapping of the variables. In the plot, query points are
indicated with colour-coded crosses and the 8 nearest neighbours of each
point are shaded in the same colour. By default `nnt()` uses the
function `RANN::nn2()` (based on the L2 metric) to find the nearest
neighbours.

``` r
library(donut)
set.seed(20092019)
x1 <- runif(100, 0, 2 * pi)
x2 <- runif(100, 0, 3)
DATA <- data.frame(x1, x2)
ranges <- rbind(c(0, 2 * pi), c(0, 3))
query <- rbind(c(6, 1.3), c(2 * pi, 3), c(3, 1.5), c(4, 0))
library(RANN)
#> Warning: package 'RANN' was built under R version 4.2.3
res2 <- nnt(DATA, query, k = 8, torus = 1:2, ranges = ranges)
plot(res2)
```

![](man/figures/README-example-1.png)<!-- -->

### Installation

To get the current released version from CRAN:

``` r
install.packages("donut")
```

### Vignette

See `vignette("donut-vignette", package = "donut")` for an overview of
the package.


NEWS.md
# donut 1.0.3

## Bug fixes and minor improvements

* Create the help file for the package correctly, with alias donut-package.

* Activated 3rd edition of the `testthat` package

# donut 1.0.2

## Bug fixes and minor improvements

* `LazyData: TRUE` has been removed from DESCRIPTION, because there is no data/ directory in donut.

# donut 1.0.1

## Dependencies

* `donut` no longer depends on the `RANN.L1` package, which has been archived on CRAN.




tests/testthat.R
library(testthat)
library(donut)

test_check("donut")


tests/testthat/test-2D.R
#context("2D check by hand")

# Check that RANN is available
got_RANN <- requireNamespace("RANN", quietly = TRUE)

# Consider cases where method 1 (on its own) gives the wrong nearest neighbours
# because an observation, or pair of observations, contribute more than once.
# Check that an additional call using method 2 inside nnt() corrects this.

if (got_RANN) {
  x1 <- c(5, 5)
  x2 <- c(9, 3)
  x3 <- c(8, 9)
  x4 <- c(0, 2)
  x5 <- c(1, 10)
  x <- rbind(x1, x2, x3, x4, x5)
#  plot(x, xlim = c(0, 10), ylim = c(0, 10))

  dfn <- function(x, y) sqrt(sum((x - y) ^ 2))

  # Only wrap on variable 1
  n1d <- n1n <- matrix(NA, 5, 5)
  temp <- apply(cbind(x1, x2, x3, x4, x5), 2, dfn, y = x1)
  n1d[1, ] <- sort(temp)
  n1n[1, ] <- order(temp)
  x4d <- c(10, 2)
  x5d <- c(11, 10)
  temp <- apply(cbind(x1, x2, x3, x4d, x5d), 2, dfn, y = x2)
  n1d[2, ] <- sort(temp)
  n1n[2, ] <- order(temp)
  temp <- apply(cbind(x1, x2, x3, x4d, x5d), 2, dfn, y = x3)
  n1d[3, ] <- sort(temp)
  n1n[3, ] <- order(temp)
  x2d <- c(-1, 3)
  x3d <- c(-2, 9)
  temp <- apply(cbind(x1, x2d, x3d, x4, x5), 2, dfn, y = x4)
  n1d[4, ] <- sort(temp)
  n1n[4, ] <- order(temp)
  temp <- apply(cbind(x1, x2d, x3d, x4, x5), 2, dfn, y = x5)
  n1d[5, ] <- sort(temp)
  n1n[5, ] <- order(temp)
  res1 <- nnt(x, x, torus = 1, ranges = c(0, 10), method = 1)
  res2 <- nnt(x, x, torus = 1, ranges = c(0, 10), method = 2)
  test_that("Wrap on variable 1: indices vs method 1", {
    testthat::expect_equal(res1$nn.idx, n1n)
  })
  test_that("Wrap on variable 1: indices vs method 2", {
    testthat::expect_equal(res2$nn.idx, n1n)
  })
  test_that("Wrap on variable 1: distances vs method 1", {
    testthat::expect_equal(res1$nn.dists, n1d)
  })
  test_that("Wrap on variable 1: distances vs method 2", {
    testthat::expect_equal(res2$nn.dists, n1d)
  })

  # Only wrap on variable 2
  n2d <- n2n <- matrix(NA, 5, 5)
  temp <- apply(cbind(x1, x2, x3, x4, x5), 2, dfn, y = x1)
  n2d[1, ] <- sort(temp)
  n2n[1, ] <- order(temp)
  x3d <- c(8, -1)
  x5d <- c(1, 0)
  temp <- apply(cbind(x1, x2, x3d, x4, x5d), 2, dfn, y = x2)
  n2d[2, ] <- sort(temp)
  n2n[2, ] <- order(temp)
  x2d <- c(9, 13)
  x4d <- c(0, 12)
  temp <- apply(cbind(x1, x2d, x3, x4d, x5), 2, dfn, y = x3)
  n2d[3, ] <- sort(temp)
  n2n[3, ] <- order(temp)
  x3d <- c(8, -1)
  x5d <- c(1, 0)
  temp <- apply(cbind(x1, x2, x3d, x4, x5d), 2, dfn, y = x4)
  n2d[4, ] <- sort(temp)
  n2n[4, ] <- order(temp)
  temp <- apply(cbind(x1, x2d, x3, x4d, x5), 2, dfn, y = x5)
  n2d[5, ] <- sort(temp)
  n2n[5, ] <- order(temp)
  res1 <- nnt(x, x, torus = 2, ranges = c(0, 10), method = 1)
  res2 <- nnt(x, x, torus = 2, ranges = c(0, 10), method = 2)
  test_that("Wrap on variable 2: indices vs method 1", {
    testthat::expect_equal(res1$nn.idx, n2n)
  })
  test_that("Wrap on variable 2: indices vs method 2", {
    testthat::expect_equal(res2$nn.idx, n2n)
  })
  test_that("Wrap on variable 2: distances vs method 1", {
    testthat::expect_equal(res1$nn.dists, n2d)
  })
  test_that("Wrap on variable 2: distances vs method 2", {
    testthat::expect_equal(res2$nn.dists, n2d)
  })

  # Wrap on variables 1 and 2
  n2d <- n2n <- matrix(NA, 5, 5)
  temp <- apply(cbind(x1, x2, x3, x4, x5), 2, dfn, y = x1)
  n2d[1, ] <- sort(temp)
  n2n[1, ] <- order(temp)
  x3d <- c(8, -1)
  x4d <- c(10, 2)
  x5d <- c(11, 0)
  temp <- apply(cbind(x1, x2, x3d, x4d, x5d), 2, dfn, y = x2)
  n2d[2, ] <- sort(temp)
  n2n[2, ] <- order(temp)
  x2d <- c(9, 13)
  x4d <- c(10, 12)
  x5d <- c(11, 10)
  temp <- apply(cbind(x1, x2d, x3, x4d, x5d), 2, dfn, y = x3)
  n2d[3, ] <- sort(temp)
  n2n[3, ] <- order(temp)
  x2d <- c(-1, 3)
  x3d <- c(-2, -1)
  x5d <- c(1, 0)
  temp <- apply(cbind(x1, x2d, x3d, x4, x5d), 2, dfn, y = x4)
  n2d[4, ] <- sort(temp)
  n2n[4, ] <- order(temp)
  x2d <- c(-1, 13)
  x3d <- c(-2, 9)
  x4d <- c(0, 12)
  temp <- apply(cbind(x1, x2d, x3d, x4d, x5), 2, dfn, y = x5)
  n2d[5, ] <- sort(temp)
  n2n[5, ] <- order(temp)
  ranges <- rbind(c(0, 10), c(0, 10))
  res1 <- nnt(x, x, torus = 1:2, ranges = ranges, method = 1)
  res2 <- nnt(x, x, torus = 1:2, ranges = ranges, method = 2)
  test_that("Wrap on variable 2: indices vs method 1", {
    testthat::expect_equal(res1$nn.idx, n2n)
  })
  test_that("Wrap on variable 2: indices vs method 2", {
    testthat::expect_equal(res2$nn.idx, n2n)
  })
  test_that("Wrap on variable 2: distances vs method 1", {
    testthat::expect_equal(res1$nn.dists, n2d)
  })
  test_that("Wrap on variable 2: distances vs method 2", {
    testthat::expect_equal(res2$nn.dists, n2d)
  })
}



tests/testthat/test-mop.R
#context("Mop up")

# Check that RANN is available
got_RANN <- requireNamespace("RANN", quietly = TRUE)

if (got_RANN) {
  set.seed(29092019)
  x1 <- runif(100, 0, 2 * pi)
  x2 <- runif(100, 0, 3)
  x3 <- runif(100, -1, 1)

  # Call nnt() with torus but missing ranges
  x <- cbind(x1, x2, x3)
  test_that("nnt: no ranges", {
    the_error <- "ranges must be supplied"
    testthat::expect_error(nnt(x, x, torus = 1, method = 1), the_error)
  })

  # Call nnt() with non-numeric torus
  test_that("nnt: non-numeric torus", {
    the_error <- "'torus' must be a non-empty numeric vector"
    testthat::expect_error(nnt(x, x, torus = "1", ranges = c(0, 1),
                               method = 1), the_error)
  })

  # Call nnt() with a torus that is not in {1, ..., ncol(data)}
  test_that("nnt: torus value inappropriate", {
    testthat::expect_error(nnt(x, x, torus = c(1:2, 5), ranges = c(0, 1),
                               method = 1))
  })

  # Call nnt() with a ranges that is not consistent with torus
  test_that("nnt: ranges inconsistent with torus", {
    testthat::expect_error(nnt(x, x, torus = 1:2, ranges = c(0, 2 * pi),
                               method = 1))
  })

  # Call nnt() with an inappropriate value of method
  test_that("nnt: method inappropriate", {
    the_error <- "method must be equal to 1 or 2"
    testthat::expect_error(nnt(x, x, torus = 1, ranges = c(0, 2 * pi),
                               method = 3), the_error)
  })

  # Call plot.nnt() with an inappropriate object
  test_that("plot.nnt: inappropriate object", {
    the_error <- "use only with \"donut\" objects"
    testthat::expect_error(plot.nnt(x), the_error)
  })

  # Call nnt() with data that are outside a range in ranges
  test_that("nnt: data out of range", {
    testthat::expect_error(nnt(x, x, torus = 1, ranges = c(0, 1)))
  })

  # Call nnt() with data that are outside a range in ranges
  query <- 7
  test_that("nnt: data out of range", {
    testthat::expect_error(nnt(x, query, torus = 1, ranges = c(0, 2 * pi)))
  })
}


R/donut.R
#' donut: Nearest Neighbour Search with Variables on a Torus
#'
#' Finds the \code{k} nearest neighbours in a dataset of specified points,
#' adding the option to wrap certain variables on a torus.  The user chooses
#' the algorithm to use to find the nearest neighbours.
#'
#' @details The function \code{\link{nnt}} performs the nearest neighbour
#'   search.  There is also a rudimentary plot method: \code{\link{plot.nnt}}.
#'
#'   The default algorithm is that provided by the function
#'   \code{\link[RANN]{nn2}} in the \code{\link[RANN]{RANN-package}}.
#'   Another possibility is the \code{\link[nabor]{knn}} function in the
#'   \code{\link[nabor]{nabor-package}}.
#'
#'   See \code{vignette("donut-vignette", package = "donut")} for an
#'   overview of the package.
#' @references Arya, S., Mount, D., Kemp, S. E. and Jefferis, G. (2019)
#'   RANN: Fast Nearest Neighbour Search (Wraps ANN Library) Using L2
#'   Metric. R package version 2.6.1.
#'   \url{https://CRAN.R-project.org/package=RANN}
#' @references Elseberg J., Magnenat S., Siegwart R., Nuchter, A. (2012)
#'   Comparison of nearest-neighbor-search strategies and implementations for
#'   efficient shape registration. \emph{Journal of Software Engineering for
#'   Robotics (JOSER)}, \strong{3}(1), 2-12
#'   \url{https://CRAN.R-project.org/package=nabor}
#' @seealso \code{\link{nnt}} for nearest neighbour with some variables
#'   wrapped on a torus.
#' @seealso \code{\link{plot.nnt}} plot method for objects returned from
#'   \code{\link{nnt}} (1 and 2 dimensional data only).
#' @docType package
#' @name donut
NULL


R/donut-package.R
#' donut: Nearest Neighbour Search with Variables on a Torus
#'
#' Finds the \code{k} nearest neighbours in a dataset of specified points,
#' adding the option to wrap certain variables on a torus.  The user chooses
#' the algorithm to use to find the nearest neighbours.
#'
#' @details The function \code{\link{nnt}} performs the nearest neighbour
#'   search.  There is also a rudimentary plot method: \code{\link{plot.nnt}}.
#'
#'   The default algorithm is that provided by the function
#'   \code{\link[RANN]{nn2}} in the \code{\link[RANN]{RANN-package}}.
#'   Another possibility is the \code{\link[nabor]{knn}} function in the
#'   \code{\link[nabor]{nabor-package}}.
#'
#'   See \code{vignette("donut-vignette", package = "donut")} for an
#'   overview of the package.
#' @references Arya, S., Mount, D., Kemp, S. E. and Jefferis, G. (2019)
#'   RANN: Fast Nearest Neighbour Search (Wraps ANN Library) Using L2
#'   Metric. R package version 2.6.1.
#'   \url{https://CRAN.R-project.org/package=RANN}
#' @references Elseberg J., Magnenat S., Siegwart R., Nuchter, A. (2012)
#'   Comparison of nearest-neighbor-search strategies and implementations for
#'   efficient shape registration. \emph{Journal of Software Engineering for
#'   Robotics (JOSER)}, \strong{3}(1), 2-12
#'   \url{https://CRAN.R-project.org/package=nabor}
#' @seealso \code{\link{nnt}} for nearest neighbour with some variables
#'   wrapped on a torus.
#' @seealso \code{\link{plot.nnt}} plot method for objects returned from
#'   \code{\link{nnt}} (1 and 2 dimensional data only).
#' @docType package
"_PACKAGE"


R/donut-internal.R
#' Internal donut functions
#'
#' Internal donut functions
#' @details
#' These functions are not intended to be called by the user.
#' @name donut-internal
#' @keywords internal
NULL

# ============================ Function for method 1 ======================== #

#' @keywords internal
#' @rdname donut-internal
method1_function <- function(data, query, k, torus, ranges, fn, ...) {
  # The number of variables to be wrapped
  nt <- length(torus)
  # The respective ranges of each of these variables
  diffs <- apply(ranges, 1, diff)
  # 3^nt factorial design
  x <- fac3(nt)
  # Multiply the -1s, 0s and 1s by the relevant values of diffs to surround the
  # original data by replicates in all directions
  x <- sweep(x, 2, diffs, "*")
  # Midpoints of each range
  mids <- rowMeans(ranges)
  # Function to replicate the data around the original data
  # Only use the variables to be wrapped
  # Each i corresponds to a different replicate
  tdata <- data[, torus, drop = FALSE]
  myfn <- function(i) {
    # Figure out which rows of tdata should be included
    # If x[i, j] < 0 then tdata[i, j] must be > mids[j]
    # If x[i, j] > 0 then tdata[i, j] must be < mids[j]
    # If x[i, j] = 0 then tdata[i, j] is unconstrained
    sgn <- sign(x[i, ])
    data_comp <- sweep(tdata, 2, sgn, "*")
    comp <- ifelse(sgn == 0, Inf, sgn * mids)
    cond <- sweep(data_comp, 2, comp, "<")
    which_rows <- apply(cond, 1, all)
    # Create the replicated data, add the other variables (if any)
    subdata <- tdata[which_rows, , drop = FALSE]
    subdata <- sweep(subdata, 2, x[i, ], "+")
    restdata <- data[which_rows, -torus, drop = FALSE]
    # Add the row numbers, so that we can return the original indices later
    idx <- which(which_rows)
    return(list(subdata = subdata, restdata = restdata, idx = idx))
  }
  # Return a list containing each replicated dataset (and the original)
  # and the indices of original data for each observation in rep_data
  # rep_data is an nrow(data)*(1+3^(nt-1)) by nt by matrix
  res <- vapply(1:nrow(x), myfn, list(subdata = 0, idx = 0, restdata = 0))
  rep_data <- do.call(rbind, res[1, ])
  rest_data <- do.call(rbind, res[2, ])
  # Replicate data
  big_data <- matrix(NA, nrow(rep_data), ncol(data), byrow = TRUE)
  big_data[, torus] <- rep_data
  big_data[, -torus] <- rest_data
  # Do the search using rep_data and the original query values
  nnres <- fn(data = big_data, query = query, k = k, ...)
  # Return to the indices that relate to the orginal data
  idx <- do.call(c, res[3, ])
  nnres$nn.idx <- apply(nnres$nn.idx, 1:2, function(x) idx[x])
  # Check for repeated indices in the rows of nnres$nn.idx
  dups <- apply(nnres$nn.idx, 1, anyDuplicated)
  # If any row has duplicated indices then call method2_function() for
  # the rows in query for which this is true
  if (any(dups > 0)) {
    which_dup <- which(dups > 0)
    res <- method2_function(data, query[which_dup, , drop = FALSE], k, torus,
                            ranges, fn, ...)
    nnres$nn.idx[which_dup, ] <- res$nn.idx
    nnres$nn.dists[which_dup, ] <- res$nn.dists
  }
  return(nnres)
}

# ============================ Function for method 2 ======================== #

#' @keywords internal
#' @rdname donut-internal
method2_function <- function(data, query, k, torus, ranges, fn, ...) {
  # Midpoints, ranges, lower and upper limits variables to be wrapped
  mids <- rowMeans(ranges)
  diffs <- apply(ranges, 1, diff)
  low <- ranges[, 1]
  high <- ranges[, 2]
  # Function to find the nearest neighbours for the jth row of query, that is,
  # for the jth point of interest
  by_query <- function(j) {
    # Take a copy of the data and shift data for the variables to be wrapped
    # so that the point of interest is at the midpoint of the each variable
    ndata <- data
    shoof <- query[j, torus] - mids
    ndata[, torus] <- sweep(ndata[, torus, drop = FALSE], 2, shoof, "-")
    # The consequence of the shift is that data are moved below or above the
    # ranges of the variables. Wrap these data round to the parts of the
    # ranges that have been vacated data that have been shifted up (shoof is
    # negative) or down (shoof is positive).
    #
    # The amounts that we need to move these data are stored in diffs.
    # The direction that we move them depends on the sign of shoof (sgn):
    # we will add sgn * diffs to such data
    sgn <- sign(shoof)
    # Multiply data for that were shifted up by -1.  Then we can identify
    # values that lie above the range by whether they are less than high
    data_comp <- sweep(ndata[, torus, drop = FALSE], 2, sgn, "*")
    # If sgn =  1: data were shifted down and values < low are shifted up
    # If sgn = -1: data were shifted up and -values < -high are shifted down
    comp <- (sgn == 1) * low - (sgn == -1) * high
    # Find the values to move (for whom comp = TRUE) and then move (only) them
    cond <- sweep(data_comp, 2, comp, "<")
    ndata[, torus] <- ndata[, torus] + sweep(cond, 2, sgn * diffs, "*")
    # Move the point of interest to the midpoint
    nquery <- query[j, , drop = FALSE]
    nquery[, torus] <- mids
    # Do the search using the new data and the new query values
    nnres <- fn(data = ndata, query = nquery, k = k, ...)
    return(nnres)
  }
  # Call by_query() for each
  res <- vapply(1:nrow(query), by_query, list(nn.idx = 0, nn.dists = 0))
  nn.idx <- do.call(rbind, res[1, ])
  nn.dists <- do.call(rbind, res[2, ])
  res <- list(nn.idx = nn.idx, nn.dists = nn.dists)
  return(res)
}

# ============================ 3^d factorial design ======================== #

#' @keywords internal
#' @rdname donut-internal
fac3 <- function(d) {
  # Creates a design matrix for a 3^d full factorial design
  #
  # Args:
  #   d : a positive integer scalar
  #
  # Returns:
  #   a 3^d by d integer matrix containing -1, 0, 1 for the levels of the d
  #   factors.  The first column varies fastest, the second column second
  #   fastest etc.
  rep_fn <- function(x) {
    rep(-1L:1L, times = 3L ^ x, each = 3L ^ (d - 1L - x))
  }
  return(vapply((d - 1L):0, rep_fn, numeric(3L ^ d)))
}


R/nearest_neighbours_on_a_torus.R
# ================================== nnt ==================================== #

#' Nearest Neighbour Search with Variables on a Torus
#'
#' Uses a user-supplied function to find the \code{k} nearest neighbours of
#' specified points in a dataset, adding the option to wrap certain variables
#' on a torus.
#'
#' @param data An \eqn{M} by \eqn{d} numeric matrix or data frame.  Each of the
#'   \eqn{M} rows contains a \eqn{d}-dimensional observation.
#' @param query An \eqn{N} by \eqn{d} numeric matrix or data frame.  Each row
#'   contains an \eqn{d}-dimensional point that will be queried against
#'   \code{data}.
#' @param k An integer scalar.  The number of nearest neighbours, of the
#'   points in the rows of \code{query}, to find.
#' @param fn The function with which to calculate the nearest neighbours.
#'   The syntax of this function must be \code{fn(data, query, k, ...)}.
#'   The default is \code{RANN::nn2}.  Another possibility is
#'   \code{nabor::knn}.
#' @param torus An integer vector with element(s) in
#'   \{1, ..., \code{ncol(data)}\}.  The corresponding variables are wrapped
#'   on the corresponding range gives in \code{ranges}.
#' @param ranges A \code{length(torus)} by \code{2} numeric matrix.
#'   Row \code{i} gives the range of variation of the variable indexed by
#'   \code{torus[i]}. \code{ranges[i, 1]} and \code{ranges[i, 2]}
#'   are equivalent values of the variable, such as 0 degrees and 360 degrees.
#'   If \code{length(torus)} = 1 then \code{ranges} may be a vector of length
#'   2.
#' @param method An integer scalar, equal to 1 or 2.  See \strong{Details}.
#' @param ... Further arguments to be passed to \code{fn}.
#' @details
#'   If \code{method = 1} then the data are partially replicated, arranged
#'   around the original data in a way that wraps the variables in \code{torus} on their respective
#'   ranges in \code{ranges}.  Then \code{fn} is called using this replicated
#'   dataset as the argument \code{data}.  If \code{k} is large and/or
#'   \code{data} is a sparse dataset then it is possible that a single
#'   observation contributes more than once to a set of nearest neighbours,
#'   which is incorrect. If this occurs then \code{nnt} uses method 2 to
#'   correct the offending rows in \code{nn.idx} and \code{nn.dists} in the
#'   returned list object.
#'
#'   If \code{method = 2} then the
#'   following approach is used for the point in each row in \code{query}.
#'   The data indexed by \code{torus} are shifted (and wrapped) so that the
#'   point is located at the respective midpoints of \code{ranges}.
#'   Method 2 is efficient only if the number of points in \code{query} is
#'   small.
#'
#'   If \code{torus} is missing then \code{fn} is called using
#'   \code{fn(data = data, query = query, k = k, ...)}, so that a call to
#'   \code{nnt} is equivalent to a call to the function chosen by \code{fn}.
#' @return An object (a list) of class \code{c("nnt", "donut")} containing the
#'   following components.
#'   \item{nn.idx}{An \eqn{N} by \eqn{d} integer matrix of the \code{k}
#'     nearest neighbour indices, i.e. the rows of \code{data}.}
#'   \item{nn.dists}{An \eqn{N} by \eqn{d} numeric matrix of the \code{k}
#'     nearest neighbour distances.}
#'   \item{data, query, k, fn}{The arguments \code{data}, \code{query},
#'     \code{k} and \code{fn} (in fact \code{substitute(fn)}).}
#'   \item{torus, ranges, method}{If \code{torus} is supplied, the
#'     arguments \code{torus}, \code{ranges} and \code{method}.}
#'   \item{call}{The call to \code{spm}.}
#' @seealso \code{\link[RANN:nn2]{RANN::nn2}},
#'   \code{\link[nabor:knn]{nabor::knn}}: nearest neighbour searches.
#' @references Arya, S., Mount, D., Kemp, S. E. and Jefferis, G. (2019)
#'   RANN: Fast Nearest Neighbour Search (Wraps ANN Library) Using L2
#'   Metric. R package version 2.6.1.
#'   \url{https://CRAN.R-project.org/package=RANN}
#' @references Elseberg J., Magnenat S., Siegwart R., Nuchter, A. (2012)
#'   Comparison of nearest-neighbor-search strategies and implementations for
#'   efficient shape registration. \emph{Journal of Software Engineering for
#'   Robotics (JOSER)}, \strong{3}(1), 2-12
#'   \url{https://CRAN.R-project.org/package=nabor}
#' @seealso \code{\link{plot.nnt}} plot method for objects returned from
#'   \code{\link{nnt}} (1 and 2 dimensional data only).
#' @examples
#' got_RANN <- requireNamespace("RANN", quietly = TRUE)
#' got_nabor <- requireNamespace("nabor", quietly = TRUE)
#'
#' set.seed(20092019)
#' # 2D example from the RANN:nn2 documentation (L2 metric)
#' x1 <- runif(100, 0, 2 * pi)
#' x2 <- runif(100, 0, 3)
#' DATA <- data.frame(x1, x2)
#' if (got_RANN) {
#'   nearest <- nnt(DATA, DATA)
#' }
#'
#' # Suppose that x1 should be wrapped
#' ranges1 <- c(0, 2 * pi)
#' query1 <- rbind(c(6, 1.3), c(2 * pi, 3), c(3, 1.5), c(4, 0))
#' if (got_RANN) {
#'   res1 <- nnt(DATA, query1, k = 8, torus = 1, ranges = ranges1)
#'   plot(res1, ylim = c(0, 3))
#' }
#'
#' # Suppose that x1 and x2 should be wrapped
#' ranges2 <- rbind(c(0, 2 * pi), c(0, 3))
#' query2 <- rbind(c(6, 1.3), c(2 * pi, 3), c(3, 1.5), c(4, 0))
#' if (got_RANN) {
#'   res2 <- nnt(DATA, query2, k = 8, torus = 1:2, ranges = ranges2)
#'   plot(res2)
#' }
#'
#' # Use nabor::knn (L2 metric) instead of RANN::nn2
#' if (got_nabor) {
#'   res3 <- nnt(DATA, query2, k = 8, fn = nabor::knn, torus = 1:2,
#'               ranges = ranges2)
#'   plot(res3)
#' }
#'
#' # 1D example
#' ranges <- c(0, 2 * pi)
#' query <- c(4, 0.1)
#' if (got_RANN) {
#'   res <- nnt(x1, query, torus = 1, ranges = ranges, method = 1)
#'   plot(res)
#' }
#' @export
nnt <- function(data, query = data, k = min(10, nrow(data)),
                fn = RANN::nn2, torus, ranges, method = 1, ...) {
  Call <- match.call(expand.dots = TRUE)
  # Make data and query matrices
  data <- as.matrix(data)
  query <- as.matrix(query)
  # Do the search and add data and query to the returned object
  if (missing(torus)) {
    res <- fn(data = data, query = query, k = k, ...)
    res <- c(res, list(data = data, query = query, k = k, fn = substitute(fn),
                       call = Call))
    class(res) <- c("nnt", "donut")
    return(res)
  }
  if (missing(ranges)) {
    stop("ranges must be supplied")
  }
  if (length(torus) == 0L || mode(torus) != "numeric") {
    stop("'torus' must be a non-empty numeric vector")
  }
  if (!all(torus %in% 1:ncol(data))) {
    stop("All elements of 'torus' must be in 1:ncol(data)")
  }
  dim_ranges <- ncol(ranges)
  if (is.null(dim_ranges)) {
    dim(ranges) <- 1:2
  }
  if (!is.matrix(ranges) || ncol(ranges) != 2 ||
      nrow(ranges) != length(torus)) {
    stop("ranges not consistent with length(torus)")
  }
  # Sort the values in the rows of ranges, from smallest to largest
  ranges <- t(apply(ranges, 1, sort))
  # Check that all data and query are in the appropriate ranges
  check_ranges <- function(i, x) {
    return(any(x[, torus[i]] < ranges[i, 1] | x[, torus[i]] > ranges[i, 2]))
  }
  check_data <- vapply(1:length(torus), check_ranges, TRUE, x = data)
  check_query <- vapply(1:length(torus), check_ranges, TRUE, x = query)
  if (any(check_data)) {
    stop("value(s) in 'data' are outside the corresponding range in 'ranges'")
  }
  if (any(check_query)) {
    stop("value(s) in 'query' are outside the corresponding range in 'ranges'")
  }
  # Check method
  if (!(method %in% 1:2)) {
    stop("method must be equal to 1 or 2")
  }
  if (method == 1) {
    res <- method1_function(data, query, k, torus, ranges, fn, ...)
  } else {
    res <- method2_function(data, query, k, torus, ranges, fn, ...)
  }
  res <- c(res, list(data = data, query = query, k = k, fn = substitute(fn),
                     torus = torus, ranges = ranges, method = method,
                     call = Call))
  class(res) <- c("nnt", "donut")
  return(res)
}

# =========================== Plot nearest neighbours ======================= #

#' Plot diagnostics for an nnt object
#'
#' \code{plot} method for an object of class \code{c("nnt")}.
#'
#' @param x an object of class \code{c("nnt")}, a result of
#'   a call to \code{\link{nnt}}.
#' @param ... Further arguments to be passed to
#'   \code{\link[graphics:plot.default]{plot}}, or
#'   \code{\link[graphics]{points}}.
#' @details This function is only applicable in 1 or 2 dimensions, that is,
#'   when \code{ncol(x$data)} = 1 or 2.  It provides a visual check that the
#'   wrapping of variables is working as intended, in cases where the
#'   number of query points, that is, \code{nrow(x$query)} is small
#'   enough that sets of nearest neighbours do not overlap much.
#'
#'   If \code{ncol(x$data)} = 1 then the index of each observation is plotted
#'   against its value, using a plotting character \code{pch = 1}.  A vertical
#'   line is superimposed at each value in \code{x$query} and the \code{x$k$}
#'   nearest neighbours of each line are colour-coded.
#'
#'   If \code{ncol(x$data)} = 2 then \code{x$data[, 2]} is plotted against
#'   \code{x$data[, 1]}, using a plotting character \code{pch = 1}.  Each point
#'   in \code{x$query} is plotted with a cross and the \code{x$k$}
#'   nearest neighbours of each point are colour-coded.
#'
#'   Colours of the lines/crosses and nearest neighbour points can be set sing an
#'   argument \code{col}.  If a variable is wrapped then the default plotting
#'   limits are set using the corresponding values in \code{x$ranges}.
#' @return Nothing is returned.
#' @seealso \code{\link{nnt}} for nearest neighbour with some variables
#'   wrapped on a torus.
#' @section Examples:
#' See the examples in \code{\link{nnt}}.
#' @export
plot.nnt <- function(x, ...) {
  if (!inherits(x, "donut")) {
    stop("use only with \"donut\" objects")
  }
  ncov <- ncol(x$data)
  if (ncov > 2) {
    stop("The plot method works for up to 2 covariates only")
  }
  if (is.null(colnames(x$data))) {
    colnames(x$data) <- paste0("X", 1:ncol(x$data))
  }
  my_plot <- function(x, ..., pch, lwd, col, xlim = my_xlim, ylim = my_ylim) {
    graphics::plot(x, ..., lwd = 1, col = "black", xlim = xlim, ylim = ylim)
  }
  my_points <- function(x, ..., pch = 16, col = "red", lwd) {
    graphics::points(x, ..., pch = pch, col = col, lwd = 1)
  }
  user_args <- list(...)
  nquery <- nrow(x$query)
  if (is.null(user_args$col)) {
    user_args$col <- 1 + 1:nquery
  }
  my_xlim <- my_ylim <- NULL
  if (ncov == 2) {
    if (!is.null(x$torus)) {
      if (1 %in% x$torus) {
        my_xlim <- x$ranges[1, ]
      }
      if (2 %in% x$torus) {
        my_ylim <- x$ranges[2, ]
      }
    }
    my_plot(x$data, ...)
    for (i in 1:nquery) {
      i_user_args <- user_args
      i_user_args$col <- user_args$col[i]
      for_my_points <- c(list(x = x$data[x$nn.idx[i, ], , drop = FALSE]),
                         i_user_args)
      do.call(my_points, for_my_points)
    }
    for_points <- list(x = x$query, col = user_args$col, pch = "x")
    do.call(graphics::points, for_points)
  } else {
    if (!is.null(x$torus)) {
      my_xlim <- x$ranges
    }
    plot_data <- cbind(x$data, index = 1:nrow(x$data))
    my_plot(plot_data, ...)
    graphics::abline(v = x$query, col = user_args$col)
    for (i in 1:nquery) {
      i_user_args <- user_args
      i_user_args$col <- user_args$col[i]
      plot_data <- cbind(x$data[x$nn.idx[i, ], , drop = FALSE], x$nn.idx[i, ])
      for_my_points <- c(list(x = plot_data), i_user_args)
      do.call(my_points, for_my_points)
    }
  }
  return(invisible())
}
